### PR TITLE
LCNV-33146 : Geofence CRUD

### PR DESCRIPTION
**Pull request by nishantbansal1910**
**WHAT**
- fetch polygon, create polygon, update polygon api.

JIRA : https://babatrucks.atlassian.net/browse/LCNV-33146

### CODE REVIEW

File: app/controllers/api/v1/polygons_controller.rb [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+                   end
+        render json: response, status: status_code
+      end
+
+      private def build_pagination
+        {
+          page: params[:page].present? ? params[:page].to_i : 1,
+          per_page: params[:per_page].present? ? params[:per_page].to_i : 10,
+        }
+      end
```
Comment (gaurav4ever) [inline app/controllers/api/v1/polygons_controller.rb:63]: Use from utilhelper

File: app/services/polygon/query_service.rb [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+      success, response = Linehaul::PolygonService.new(current_user["authentication_token"]).fetch_polygon_details(pagination, params[:name], params[:active])
+
+      handle_errors(response) && return unless success
+
+      if success && response["data"] && response["total_count"]
+        pagination[:total_count] = response["total_count"]
+        response["pagination"] = pagination
+        response
+      else
+        handle_errors("Techincal_issue")
```
Comment (gaurav4ever) [inline app/services/polygon/query_service.rb:28]: This will not match with switch case

File: app/services/linehaul/polygon_service.rb [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+    POLYGON_URL = LINEHAUL_BASE_URL + "/api/v2/polygon"
+
+    attr_accessor :auth_token
+
+    def initialize(auth_token)
+      self.auth_token = auth_token
+    end
+
+    def fetch_polygon_details(pagination, name, active)
+      start_index = pagination[:page] * pagination[:per_page]
```
Comment (gaurav4ever) [inline app/services/linehaul/polygon_service.rb:19]: `.to_i` on every math functions

File: app/services/linehaul/polygon_service.rb [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+      end_index = (pagination[:page] + 1) * pagination[:per_page]
+      response = Typhoeus::Request.new(
+        FETCH_POLYGONS_URL + "?start_index=" + start_index.to_s + "&end_index=" + end_index.to_s + "&filter=" + name.to_s + "&active=" + active.to_s,
+        headers: {
+          "Authorization": auth_token,
```
Comment (gaurav4ever) [inline app/services/linehaul/polygon_service.rb:24]: No need to double quotes

File: app/services/linehaul/polygon_service.rb [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+          response_data = JSON.parse(response.body)
+          [false, response_data&.dig("message")]
+        end
+      else
+        [false, "Technical issue"]
+      end
+    end
+
+  end
+end
```
Comment (gaurav4ever) [inline app/services/linehaul/polygon_service.rb:94]: Add a new empty line

File: app/services/polygon/creation_service.rb [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+        errors << "Technical issue, please try again later"
+        self.error_code = :technical_issue
+      else
+        errors << error_response
+        self.error_code = :something_went_wrong
+      end
+    end
+
+  end
+end
```
Comment (gaurav4ever) [inline app/services/polygon/creation_service.rb:61]: Add a new empty line

File: app/services/polygon/query_service.rb [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+        errors << "Technical issue, please try again later"
+        self.error_code = :technical_issue
+      else
+        errors << error_response
+        self.error_code = :something_went_wrong
+      end
+    end
+
+  end
+end
```
Comment (gaurav4ever) [inline app/services/polygon/query_service.rb:58]: Add a new empty line

File: app/services/polygon/updation_service.rb [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+        errors << "Technical issue, please try again later"
+        self.error_code = :technical_issue
+      else
+        errors << error_response
+        self.error_code = :something_went_wrong
+      end
+    end
+
+  end
+end
```
Comment (gaurav4ever) [inline app/services/polygon/updation_service.rb:47]: Add a new empty line

File: app/controllers/api/v1/polygons_controller.rb [hunk 9]
```diff
@@ -0,0 +1,79 @@
...
+            :unprocessable_entity
+          end
+        else
+          :ok
+        end
+      end
+
+    end
+  end
+end
```
Comment (gaurav4ever) [inline app/controllers/api/v1/polygons_controller.rb:79]: Add a new empty line

File: app/services/polygon/creation_service.rb [hunk 10]
```diff
@@ -0,0 +1,61 @@
...
+      self.params = request_params
+      self.errors = []
+    end
+
+    def run!
+      validate_params
+      create_polygon if errors.empty?
+    end
+
+    def validate_params
```
Comment (gaurav4ever) [inline app/services/polygon/creation_service.rb:20]: Add validation on distance_unit as well. On `m` should be allowed 
https://github.com/loconav-tech/marketplace/blob/master/src/main/java/com/loconav/marketplace/partner/enums/DistanceUnit.java

File: app/services/polygon/updation_service.rb [hunk 11]
```diff
@@ -0,0 +1,47 @@
...
+
+    def update_polygon
+      success, response = Linehaul::PolygonService.new(current_account["authentication_token"]).update_polygon(params, polygon_id)
+
+      handle_errors(response) && return unless success
+
+      if success && response.present? && response["polygon"].present?
+        response["polygon"]
+      else
+        handle_errors("Technical issue, please try again later")
```
Comment (gaurav4ever) [inline app/services/polygon/updation_service.rb:27]: This will not match with switch case

**Review by gaurav4ever — CHANGES_REQUESTED** [review]
Add new line at the end in each new file

File: app/services/linehaul/polygon_service.rb [hunk 12]
```diff
@@ -0,0 +1,94 @@
...
+    POLYGON_URL = LINEHAUL_BASE_URL + "/api/v2/polygon"
+
+    attr_accessor :auth_token
+
+    def initialize(auth_token)
+      self.auth_token = auth_token
+    end
+
+    def fetch_polygon_details(pagination, name, active)
+      start_index = ((pagination[:page]-1) * pagination[:per_page]).to_i
```
Comment (gaurav4ever) [inline app/services/linehaul/polygon_service.rb:19]: Code not formatted.

`start_index = ((pagination[:page].to_i - 1) * pagination[:per_page].to_i)`

File: app/services/polygon/creation_service.rb [hunk 13]
```diff
@@ -0,0 +1,69 @@
...
+
+    def initialize(current_account, request_params)
+      self.current_account = current_account
+      self.params = request_params
+      self.errors = []
+    end
+
+    def run!
+      validate_params
+      create_polygon if errors.empty?
```
Comment (gaurav4ever) [inline app/services/polygon/creation_service.rb:17]: move `if errors.empty?` inside method

File: app/services/polygon/creation_service.rb [hunk 14]
```diff
@@ -0,0 +1,69 @@
...
+        self.error_code = :parameter_is_invalid
+      when "Technical issue"
+        errors << "Technical issue, please try again later"
+        self.error_code = :technical_issue
+      else
+        errors << error_response
+        self.error_code = :something_went_wrong
+      end
+    end
+
```
Comment (gaurav4ever) [inline app/services/polygon/creation_service.rb:67]: remove new line

File: app/services/polygon/creation_service.rb [hunk 15]
```diff
@@ -0,0 +1,69 @@
+module Polygon
+  class CreationService
+
+    QUERY_PARAMS = %i[name lat radius long distance_unit active]
```
Comment (gaurav4ever) [inline app/services/polygon/creation_service.rb:4]: This is a creation service. Don't use **QUERY_PARAMS** name. 
Use **PERMITTED_PARAMS**

File: app/services/polygon/creation_service.rb [hunk 16]
```diff
@@ -0,0 +1,69 @@
...
+    REQUIRED_PARAMS = %w[name lat long radius]
```
Comment (gaurav4ever) [inline app/services/polygon/creation_service.rb:5]: We can name this **MANDATORY_PARAMS**

File: app/services/polygon/updation_service.rb [hunk 17]
```diff
@@ -0,0 +1,55 @@
+module Polygon
+  class UpdationService
+
+    QUERY_PARAMS = %i[name lat radius long distance_unit active]
```
Comment (gaurav4ever) [inline app/services/polygon/updation_service.rb:4]: QUERY_PARAMS -> PERMITTED_PARAMS

File: app/controllers/api/v1/polygons_controller.rb [hunk 18]
```diff
@@ -0,0 +1,72 @@
+module Api
+  module V1
+    class PolygonsController < ApplicationController
+      include AuthenticationHelper, UtilHelper
+
+      before_action :authenticate_account
+
+      def index
```
Comment (gaurav4ever) [inline app/controllers/api/v1/polygons_controller.rb:8]: permitted params ?

File: config/secrets.yml [hunk 19]
```diff
@@ -3,8 +3,9 @@ common: &common_settings
   secret_key_base: <%= ENV.fetch("SECRET_KEY_BASE") %>
   header_user_authentication: <%= ENV.fetch("HEADER_USER_AUTHENTICATION") %>
   redis_host: <%= ENV.fetch("REDIS_HOST") %>
-  redis_db_index: <%= ENV.fetch("REDIS_DB_INDEX") || '0' %>
```
Comment (gaurav4ever) [inline config/secrets.yml:6]: revert

**Review by gaurav4ever — CHANGES_REQUESTED** [review]

File: app/services/polygon/creation_service.rb (hunk 14 above)
Comment (nishantbansal1910) [reply app/services/polygon/creation_service.rb:74]: same as another code

**Review by gaurav4ever — APPROVED** [review]

**Pull request merged**